graph/src/extension.rs

```diff
@@ -0,0 +1,120 @@
+use futures::sync::oneshot;
+use std::sync::{Arc, Mutex};
+use tokio::prelude::{Future, Poll, Stream};
+
+pub struct Cancelable<T, C> {
+    cancelable: T,
+    canceled: oneshot::Receiver<()>,
+    on_cancel: C,
+}
+
+/// It's not viable to use `select` directly, so we do a custom implementation.
+impl<S: Stream, C: Fn() -> S::Error> Stream for Cancelable<S, C> {
+    type Item = S::Item;
+    type Error = S::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        // Error if the stream was canceled, otherwise poll it.
+        if self.canceled.poll().is_err() {
+            Err((self.on_cancel)())
+        } else {
+            self.cancelable.poll()
+        }
+    }
+}
+
+impl<F: Future, C: Fn() -> F::Error> Future for Cancelable<F, C> {
+    type Item = F::Item;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        // Error if the future was canceled, otherwise poll it.
+        if self.canceled.poll().is_err() {
+            Err((self.on_cancel)())
+        } else {
+            self.cancelable.poll()
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct CancelGuard {
+    canceler: oneshot::Sender<()>,
+}
+
+impl CancelGuard {
+    /// A more readable `drop`.
+    pub fn cancel(self) {}
+
+    /// Convert into a `SharedGuard`.
+    pub fn shared(self) -> SharedCancelGuard {
+        SharedCancelGuard {
+            guard: Arc::new(Mutex::new(Some(self.canceler))),
+        }
+    }
+}
+
+/// A version of `CancelGuard` that can be cloned.
+///
+/// To cancel the stream, call `cancel` on one of the guards or drop all guards.
+#[derive(Clone, Debug)]
+pub struct SharedCancelGuard {
+    guard: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+}
+
+impl SharedCancelGuard {
+    /// Cancels the stream, a noop if already canceled.
+    pub fn cancel(&self) {
+        *self.guard.lock().unwrap() = None
+    }
+
+    pub fn is_canceled(&self) -> bool {
+        self.guard.lock().unwrap().is_some()
```

Comment:
Unless I'm reading this wrong, I think this check should be `.is_none()`.

Reply:
I had a 50% chance of getting it right.

Reply:
🤣